File: src/lib/verify/service.rs
use crate::subtitle::model::Subtitle;
use crate::verify::model::{ComparisonEntry, ComparisonStatus, VerificationReport};
use log::debug;
use std::collections::HashMap;

/// Compare two sets of subtitles and generate a verification report
///
/// The reference subtitles are considered authoritative (correct).
/// The target subtitles are verified against the reference.
pub fn compare_subtitles(
    ref_subs: Vec<Subtitle>,
    ref_file: String,
    target_subs: Vec<Subtitle>,
    target_file: String,
) -> VerificationReport {
    let total_ref_count = ref_subs.len();
    let total_target_count = target_subs.len();

    debug!(
        "comparing {} reference subtitles with {} target subtitles",
        total_ref_count, total_target_count
    );

    let _ref_map: HashMap<u32, Subtitle> = ref_subs
        .iter()
        .map(|s| (*s.index.as_ref(), s.clone()))
        .collect();

    let target_map: HashMap<u32, Subtitle> = target_subs
        .iter()
        .map(|s| (*s.index.as_ref(), s.clone()))
        .collect();

    let mut perfect_matches = 0;
    let mut timestamp_mismatches = Vec::new();
    let mut missing_in_target = Vec::new();

    debug!("phase 1: exact index matching");

    for ref_sub in &ref_subs {
        let ref_idx = *ref_sub.index.as_ref();

        if let Some(target_sub) = target_map.get(&ref_idx) {
            if timestamps_match(ref_sub, target_sub) {
                perfect_matches += 1;
            } else {
                timestamp_mismatches.push(ComparisonEntry {
                    ref_subtitle: ref_sub.clone(),
                    status: ComparisonStatus::TimestampMismatch {
                        ref_start: Subtitle::format_timestamp(ref_sub.start_time.as_ref()),
                        ref_end: Subtitle::format_timestamp(ref_sub.end_time.as_ref()),
                        target_start: Subtitle::format_timestamp(target_sub.start_time.as_ref()),
                        target_end: Subtitle::format_timestamp(target_sub.end_time.as_ref()),
                    },
                });
            }
        } else {
            missing_in_target.push(ComparisonEntry {
                ref_subtitle: ref_sub.clone(),
                status: ComparisonStatus::MissingInTarget,
            });
        }
    }

    debug!(
        "phase 1 results: {} perfect matches, {} timestamp mismatches, {} missing",
        perfect_matches,
        timestamp_mismatches.len(),
        missing_in_target.len()
    );

    let detected_offset = if !missing_in_target.is_empty() {
        debug!("phase 2: offset detection (range -10 to +10)");
        detect_index_offset(&missing_in_target, &target_map)
    } else {
        debug!("phase 2: skipping offset detection (no missing subtitles)");
        None
    };

    if let Some(offset) = detected_offset {
        debug!("detected offset: {}", offset);
    } else {
        debug!("no offset detected");
    }

    let mut matched_with_offset = Vec::new();
    let mut remaining_missing = Vec::new();

    if let Some(offset) = detected_offset {
        debug!("phase 3: reclassifying with offset {}", offset);

        for entry in missing_in_target {
            let ref_idx = *entry.ref_subtitle.index.as_ref();
            let target_idx = (ref_idx as i32 + offset) as u32;

            if let Some(target_sub) = target_map.get(&target_idx) {
                if timestamps_match(&entry.ref_subtitle, target_sub) {
                    matched_with_offset.push(ComparisonEntry {
                        ref_subtitle: entry.ref_subtitle.clone(),
                        status: ComparisonStatus::MatchedWithOffset {
                            offset,
                            target_index: target_idx,
                        },
                    });
                } else {
                    remaining_missing.push(entry);
                }
            } else {
                remaining_missing.push(entry);
            }
        }

        debug!(
            "phase 3 results: {} matched with offset, {} still missing",
            matched_with_offset.len(),
            remaining_missing.len()
        );
    } else {
        remaining_missing = missing_in_target;
    }

    debug!("phase 4: finding extra subtitles in target");

    let mut matched_target_indices: HashMap<u32, bool> = HashMap::new();

    for ref_sub in &ref_subs {
        let ref_idx = *ref_sub.index.as_ref();
        if target_map.contains_key(&ref_idx) {
            matched_target_indices.insert(ref_idx, true);
        }
    }

    for entry in &matched_with_offset {
        if let ComparisonStatus::MatchedWithOffset { target_index, .. } = entry.status {
            matched_target_indices.insert(target_index, true);
        }
    }

    let extra_in_target: Vec<Subtitle> = target_subs
        .iter()
        .filter(|s| {
            let idx = *s.index.as_ref();
            !matched_target_indices.contains_key(&idx)
        })
        .cloned()
        .collect();

    debug!("phase 4 results: {} extra in target", extra_in_target.len());

    VerificationReport {
        ref_file,
        target_file,
        total_ref_count,
        total_target_count,
        perfect_matches,
        timestamp_mismatches,
        missing_in_target: remaining_missing,
        matched_with_offset,
        detected_offset,
        extra_in_target,
    }
}

/// Check if two subtitles have matching timestamps
fn timestamps_match(s1: &Subtitle, s2: &Subtitle) -> bool {
    s1.start_time == s2.start_time && s1.end_time == s2.end_time
}

/// Detect the most common index offset between reference and target
///
/// Tests offsets from -10 to +10 and returns the most frequent offset
/// that results in timestamp matches.
fn detect_index_offset(
    missing: &[ComparisonEntry],
    target_map: &HashMap<u32, Subtitle>,
) -> Option<i32> {
    let mut offset_counts: HashMap<i32, usize> = HashMap::new();

    for entry in missing {
        let ref_idx = *entry.ref_subtitle.index.as_ref();

        for offset in -10..=10 {
            if offset == 0 {
                continue;
            }

            let target_idx = ref_idx as i32 + offset;
            if target_idx < 1 {
                continue;
            }

            if let Some(target_sub) = target_map.get(&(target_idx as u32))
                && timestamps_match(&entry.ref_subtitle, target_sub)
            {
                *offset_counts.entry(offset).or_insert(0) += 1;
            }
        }
    }

    offset_counts
        .into_iter()
        .max_by_key(|(offset, count)| (*count, -offset.abs()))
        .map(|(offset, _)| offset)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::subtitle::model::{SubtitleIndex, SubtitleText, SubtitleTimestamp};
    use chrono::Duration;

    fn make_test_subtitle(index: u32, start_ms: i64, end_ms: i64, text: &str) -> Subtitle {
        Subtitle::new(
            SubtitleIndex::try_new(index).unwrap(),
            SubtitleTimestamp::try_new(Duration::milliseconds(start_ms)).unwrap(),
            SubtitleTimestamp::try_new(Duration::milliseconds(end_ms)).unwrap(),
            SubtitleText::try_new(text.to_string()).unwrap(),
        )
        .unwrap()
    }

    #[test]
    fn test_timestamps_match() {
        let s1 = make_test_subtitle(1, 1000, 2000, "Text 1");
        let s2 = make_test_subtitle(2, 1000, 2000, "Text 2");
        let s3 = make_test_subtitle(3, 1000, 2001, "Text 3");

        assert!(timestamps_match(&s1, &s2)); // Same timestamps
        assert!(!timestamps_match(&s1, &s3)); // Different end time
    }

    #[test]
    fn test_compare_perfect_match() {
        let ref_subs = vec![
            make_test_subtitle(1, 1000, 2000, "English 1"),
            make_test_subtitle(2, 2000, 3000, "English 2"),
            make_test_subtitle(3, 3000, 4000, "English 3"),
        ];

        let target_subs = vec![
            make_test_subtitle(1, 1000, 2000, "Russian 1"),
            make_test_subtitle(2, 2000, 3000, "Russian 2"),
            make_test_subtitle(3, 3000, 4000, "Russian 3"),
        ];

        let report = compare_subtitles(
            ref_subs,
            "ref.srt".to_string(),
            target_subs,
            "target.srt".to_string(),
        );

        assert_eq!(report.perfect_matches, 3);
        assert!(report.timestamp_mismatches.is_empty());
        assert!(report.missing_in_target.is_empty());
        assert!(report.matched_with_offset.is_empty());
        assert!(report.extra_in_target.is_empty());
        assert_eq!(report.detected_offset, None);
        assert!(report.is_perfect());
        assert_eq!(report.match_percentage(), 100.0);
    }

    #[test]
    fn test_compare_missing_subtitles() {
        let ref_subs = vec![
            make_test_subtitle(1, 1000, 2000, "Text 1"),
            make_test_subtitle(2, 2000, 3000, "Text 2"),
            make_test_subtitle(3, 3000, 4000, "Text 3"),
        ];

        let target_subs = vec![make_test_subtitle(1, 1000, 2000, "Text 1")];

        let report = compare_subtitles(
            ref_subs,
            "ref.srt".to_string(),
            target_subs,
            "target.srt".to_string(),
        );

        assert_eq!(report.perfect_matches, 1);
        assert_eq!(report.missing_in_target.len(), 2);
        assert!((report.match_percentage() - 33.333333333333336).abs() < 0.0001);
    }

    #[test]
    fn test_compare_timestamp_mismatch() {
        let ref_subs = vec![
            make_test_subtitle(1, 1000, 2000, "Text 1"),
            make_test_subtitle(2, 2000, 3000, "Text 2"),
        ];

        let target_subs = vec![
            make_test_subtitle(1, 1000, 2000, "Text 1"),
            make_test_subtitle(2, 2100, 3100, "Text 2"), // Different timestamps
        ];

        let report = compare_subtitles(
            ref_subs,
            "ref.srt".to_string(),
            target_subs,
            "target.srt".to_string(),
        );

        assert_eq!(report.perfect_matches, 1);
        assert_eq!(report.timestamp_mismatches.len(), 1);
        assert_eq!(report.match_percentage(), 50.0);
    }

    #[test]
    fn test_compare_with_offset() {
        let ref_subs = vec![
            make_test_subtitle(5, 1000, 2000, "English 5"),
            make_test_subtitle(6, 2000, 3000, "English 6"),
            make_test_subtitle(7, 3000, 4000, "English 7"),
        ];

        // Target has indices offset by -2
        let target_subs = vec![
            make_test_subtitle(3, 1000, 2000, "Russian 3"),
            make_test_subtitle(4, 2000, 3000, "Russian 4"),
            make_test_subtitle(5, 3000, 4000, "Russian 5"),
        ];

        let report = compare_subtitles(
            ref_subs,
            "ref.srt".to_string(),
            target_subs,
            "target.srt".to_string(),
        );

        assert_eq!(report.perfect_matches, 0);
        assert_eq!(report.timestamp_mismatches.len(), 1); // Index 5 has different timestamps
        assert_eq!(report.matched_with_offset.len(), 2); // Indices 6 and 7 with offset -2
        assert_eq!(report.detected_offset, Some(-2));
        // Only 2 out of 3 matched (the third has timestamp mismatch)
        assert!((report.match_percentage() - 66.66666666666667).abs() < 0.0001);
        assert!(!report.is_perfect()); // Not perfect
        assert!(report.has_issues()); // Has timestamp mismatch issue
    }

    #[test]
    fn test_compare_extra_in_target() {
        let ref_subs = vec![make_test_subtitle(1, 1000, 2000, "Text 1")];

        let target_subs = vec![
            make_test_subtitle(1, 1000, 2000, "Text 1"),
            make_test_subtitle(2, 2000, 3000, "Text 2"),
            make_test_subtitle(3, 3000, 4000, "Text 3"),
        ];

        let report = compare_subtitles(
            ref_subs,
            "ref.srt".to_string(),
            target_subs,
            "target.srt".to_string(),
        );

        assert_eq!(report.perfect_matches, 1);
        assert_eq!(report.extra_in_target.len(), 2);
        assert_eq!(report.match_percentage(), 100.0);
    }

    #[test]
    fn test_detect_offset_simple() {
        let ref_subs = vec![
            make_test_subtitle(10, 1000, 2000, "Text"),
            make_test_subtitle(20, 2000, 3000, "Text"),
        ];

        let target_map: HashMap<u32, Subtitle> = vec![
            make_test_subtitle(8, 1000, 2000, "Text"),
            make_test_subtitle(18, 2000, 3000, "Text"),
        ]
        .into_iter()
        .map(|s| (*s.index.as_ref(), s))
        .collect();

        let missing: Vec<ComparisonEntry> = ref_subs
            .into_iter()
            .map(|s| ComparisonEntry {
                ref_subtitle: s,
                status: ComparisonStatus::MissingInTarget,
            })
            .collect();

        let offset = detect_index_offset(&missing, &target_map);
        assert_eq!(offset, Some(-2));
    }

    #[test]
    fn test_detect_offset_no_match() {
        let ref_subs = vec![make_test_subtitle(10, 1000, 2000, "Text")];

        let target_map: HashMap<u32, Subtitle> = vec![make_test_subtitle(8, 9000, 9999, "Text")]
            .into_iter()
            .map(|s| (*s.index.as_ref(), s))
            .collect();

        let missing: Vec<ComparisonEntry> = ref_subs
            .into_iter()
            .map(|s| ComparisonEntry {
                ref_subtitle: s,
                status: ComparisonStatus::MissingInTarget,
            })
            .collect();

        let offset = detect_index_offset(&missing, &target_map);
        assert_eq!(offset, None);
    }

    #[test]
    fn test_compare_empty_files() {
        let ref_subs: Vec<Subtitle> = vec![];
        let target_subs: Vec<Subtitle> = vec![];

        let report = compare_subtitles(
            ref_subs,
            "ref.srt".to_string(),
            target_subs,
            "target.srt".to_string(),
        );

        assert_eq!(report.total_ref_count, 0);
        assert_eq!(report.total_target_count, 0);
        assert_eq!(report.match_percentage(), 0.0);
    }
}

File: pkg/verify/service.go
package verify

import (
	"log/slog"

	"github.com/lebe-dev/submarine/pkg/subtitle"
)

// CompareSubtitles compares two sets of subtitles and generates a verification
// report. Was the free function `compare_subtitles`.
//
// The reference subtitles are considered authoritative (correct). The target
// subtitles are verified against the reference.
func CompareSubtitles(
	refSubs []subtitle.Subtitle,
	refFile string,
	targetSubs []subtitle.Subtitle,
	targetFile string,
) VerificationReport {
	totalRefCount := len(refSubs)
	totalTargetCount := len(targetSubs)

	slog.Debug("comparing reference subtitles with target subtitles",
		"ref_count", totalRefCount, "target_count", totalTargetCount)

	// _ref_map in Rust is unused (prefixed with underscore); omitted here.

	targetMap := make(map[uint32]subtitle.Subtitle, len(targetSubs))
	for _, s := range targetSubs {
		targetMap[s.Index.Value()] = s
	}

	perfectMatches := 0
	var timestampMismatches []ComparisonEntry
	var missingInTarget []ComparisonEntry

	slog.Debug("phase 1: exact index matching")

	for i := range refSubs {
		refSub := refSubs[i]
		refIdx := refSub.Index.Value()

		if targetSub, ok := targetMap[refIdx]; ok {
			if timestampsMatch(&refSub, &targetSub) {
				perfectMatches++
			} else {
				timestampMismatches = append(timestampMismatches, ComparisonEntry{
					RefSubtitle: refSub,
					Status: ComparisonStatus{
						Kind:        TimestampMismatch,
						RefStart:    subtitle.FormatTimestamp(refSub.StartTime.Value()),
						RefEnd:      subtitle.FormatTimestamp(refSub.EndTime.Value()),
						TargetStart: subtitle.FormatTimestamp(targetSub.StartTime.Value()),
						TargetEnd:   subtitle.FormatTimestamp(targetSub.EndTime.Value()),
					},
				})
			}
		} else {
			missingInTarget = append(missingInTarget, ComparisonEntry{
				RefSubtitle: refSub,
				Status:      ComparisonStatus{Kind: MissingInTarget},
			})
		}
	}

	slog.Debug("phase 1 results",
		"perfect_matches", perfectMatches,
		"timestamp_mismatches", len(timestampMismatches),
		"missing", len(missingInTarget))

	var detectedOffset *int32
	if len(missingInTarget) != 0 {
		slog.Debug("phase 2: offset detection (range -10 to +10)")
		detectedOffset = detectIndexOffset(missingInTarget, targetMap)
	} else {
		slog.Debug("phase 2: skipping offset detection (no missing subtitles)")
		detectedOffset = nil
	}

	if detectedOffset != nil {
		slog.Debug("detected offset", "offset", *detectedOffset)
	} else {
		slog.Debug("no offset detected")
	}

	var matchedWithOffset []ComparisonEntry
	var remainingMissing []ComparisonEntry

	if detectedOffset != nil {
		offset := *detectedOffset
		slog.Debug("phase 3: reclassifying with offset", "offset", offset)

		for _, entry := range missingInTarget {
			refIdx := entry.RefSubtitle.Index.Value()
			targetIdx := uint32(int32(refIdx) + offset)

			if targetSub, ok := targetMap[targetIdx]; ok {
				if timestampsMatch(&entry.RefSubtitle, &targetSub) {
					matchedWithOffset = append(matchedWithOffset, ComparisonEntry{
						RefSubtitle: entry.RefSubtitle,
						Status: ComparisonStatus{
							Kind:        MatchedWithOffset,
							Offset:      offset,
							TargetIndex: targetIdx,
						},
					})
				} else {
					remainingMissing = append(remainingMissing, entry)
				}
			} else {
				remainingMissing = append(remainingMissing, entry)
			}
		}

		slog.Debug("phase 3 results",
			"matched_with_offset", len(matchedWithOffset),
			"still_missing", len(remainingMissing))
	} else {
		remainingMissing = missingInTarget
	}

	slog.Debug("phase 4: finding extra subtitles in target")

	matchedTargetIndices := make(map[uint32]bool)

	for i := range refSubs {
		refIdx := refSubs[i].Index.Value()
		if _, ok := targetMap[refIdx]; ok {
			matchedTargetIndices[refIdx] = true
		}
	}

	for _, entry := range matchedWithOffset {
		if entry.Status.Kind == MatchedWithOffset {
			matchedTargetIndices[entry.Status.TargetIndex] = true
		}
	}

	var extraInTarget []subtitle.Subtitle
	for _, s := range targetSubs {
		idx := s.Index.Value()
		if !matchedTargetIndices[idx] {
			extraInTarget = append(extraInTarget, s)
		}
	}

	slog.Debug("phase 4 results", "extra_in_target", len(extraInTarget))

	return VerificationReport{
		RefFile:             refFile,
		TargetFile:          targetFile,
		TotalRefCount:       totalRefCount,
		TotalTargetCount:    totalTargetCount,
		PerfectMatches:      perfectMatches,
		TimestampMismatches: timestampMismatches,
		MissingInTarget:     remainingMissing,
		MatchedWithOffset:   matchedWithOffset,
		DetectedOffset:      detectedOffset,
		ExtraInTarget:       extraInTarget,
	}
}

// timestampsMatch reports whether two subtitles have matching timestamps. Was
// the private fn `timestamps_match`.
func timestampsMatch(s1, s2 *subtitle.Subtitle) bool {
	return s1.StartTime == s2.StartTime && s1.EndTime == s2.EndTime
}

// detectIndexOffset detects the most common index offset between reference and
// target. Was the private fn `detect_index_offset`.
//
// Tests offsets from -10 to +10 and returns the most frequent offset that
// results in timestamp matches.
func detectIndexOffset(
	missing []ComparisonEntry,
	targetMap map[uint32]subtitle.Subtitle,
) *int32 {
	offsetCounts := make(map[int32]int)

	for _, entry := range missing {
		refIdx := entry.RefSubtitle.Index.Value()

		for offset := int32(-10); offset <= 10; offset++ {
			if offset == 0 {
				continue
			}

			targetIdx := int32(refIdx) + offset
			if targetIdx < 1 {
				continue
			}

			if targetSub, ok := targetMap[uint32(targetIdx)]; ok && timestampsMatch(&entry.RefSubtitle, &targetSub) {
				offsetCounts[offset]++
			}
		}
	}

	// Equivalent of Rust's
	//   offset_counts.into_iter().max_by_key(|(offset, count)| (*count, -offset.abs()))
	// Key ordering: primarily by count (higher wins), then by -offset.abs()
	// (smaller absolute offset wins). HashMap iteration order is unspecified in
	// both Rust and Go, so true ties resolve non-deterministically in both.
	found := false
	var bestOffset int32
	var bestCount int
	var bestNegAbs int32
	for offset, count := range offsetCounts {
		negAbs := -absI32(offset)
		if !found || count > bestCount || (count == bestCount && negAbs > bestNegAbs) {
			found = true
			bestOffset = offset
			bestCount = count
			bestNegAbs = negAbs
		}
	}

	if !found {
		return nil
	}
	return &bestOffset
}

// absI32 returns the absolute value of an int32 (Rust i32::abs).
func absI32(v int32) int32 {
	if v < 0 {
		return -v
	}
	return v
}
